File: runtimes/node/src/reliability.ts
// Node runtime's implementation of the shared reliability spec (T023).
// The constants MUST match core/src/runtime-spec/reliability-spec.ts — the cross-language
// conformance suite (T068) asserts this. Retry is gated on idempotency so non-idempotent
// mutating calls are never retried (Constitution Principle IV).

export const RETRY_SPEC = {
  maxAttempts: 3,
  baseDelayMs: 200,
  factor: 2,
  jitterFraction: 0.5,
  transientStatuses: [408, 429, 500, 502, 503, 504],
};

export function isRetryable(params: { idempotent: boolean; status?: number; isTimeout?: boolean }): boolean {
  if (!params.idempotent) return false;
  if (params.isTimeout) return true;
  if (params.status == null) return false;
  return RETRY_SPEC.transientStatuses.includes(params.status);
}

export function baseDelayForAttempt(attempt: number): number {
  return RETRY_SPEC.baseDelayMs * Math.pow(RETRY_SPEC.factor, attempt);
}

export function backoffSchedule(): number[] {
  const waits: number[] = [];
  for (let attempt = 0; attempt < RETRY_SPEC.maxAttempts - 1; attempt++) {
    waits.push(baseDelayForAttempt(attempt));
  }
  return waits;
}

export interface ToolInvocationLogRecord {
  ts: string;
  level: "info" | "warn" | "error";
  event: "tool_invocation";
  toolName: string;
  attempts: number;
  outcome: "success" | "failure";
  durationMs: number;
  error?: string;
}

export function redactForLog(value: string): string {
  return value.replace(/(bearer\s+)[A-Za-z0-9._-]+/gi, "$1<redacted>");
}

export interface InvokeResult {
  ok: boolean;
  status?: number;
  body?: unknown;
  error?: string;
  attempts: number;
}

/** Invoke a callable with retry/backoff per the spec, emitting one structured log record.
 * `sleep` is injectable so tests don't wait real backoff delays. */
export async function invokeWithReliability(
  toolName: string,
  idempotent: boolean,
  call: (attempt: number) => Promise<{ status: number; body: unknown }>,
  opts: { log?: (r: ToolInvocationLogRecord) => void; sleep?: (ms: number) => Promise<void> } = {},
): Promise<InvokeResult> {
  const log = opts.log ?? ((r) => process.stderr.write(JSON.stringify(r) + "\n"));
  const sleep = opts.sleep ?? ((ms) => new Promise((r) => setTimeout(r, ms)));
  const started = Date.now();
  let attempts = 0;
  let lastError: string | undefined;
  let lastStatus: number | undefined;

  for (let attempt = 0; attempt < RETRY_SPEC.maxAttempts; attempt++) {
    attempts = attempt + 1;
    try {
      const res = await call(attempt);
      lastStatus = res.status;
      if (res.status < 400) {
        log({
          ts: new Date().toISOString(),
          level: "info",
          event: "tool_invocation",
          toolName,
          attempts,
          outcome: "success",
          durationMs: Date.now() - started,
        });
        return { ok: true, status: res.status, body: res.body, attempts };
      }
      lastError = `HTTP ${res.status}`;
      if (!isRetryable({ idempotent, status: res.status }) || attempt === RETRY_SPEC.maxAttempts - 1) {
        break;
      }
    } catch (e) {
      lastError = redactForLog((e as Error).message);
      if (!isRetryable({ idempotent, isTimeout: true }) || attempt === RETRY_SPEC.maxAttempts - 1) {
        break;
      }
    }
    const jitter = 1 + (Math.random() * 2 - 1) * RETRY_SPEC.jitterFraction;
    await sleep(baseDelayForAttempt(attempt) * jitter);
  }

  log({
    ts: new Date().toISOString(),
    level: "error",
    event: "tool_invocation",
    toolName,
    attempts,
    outcome: "failure",
    durationMs: Date.now() - started,
    error: lastError,
  });
  return { ok: false, status: lastStatus, error: lastError, attempts };
}

File: runtimes/python/src/wrapper_runtime/reliability.py
"""Python runtime reliability implementation of the shared spec (T023).

Constants MUST match core/src/runtime-spec/reliability-spec.ts and the Node/Java
runtimes; the cross-language conformance suite (T068) asserts this.
"""
from __future__ import annotations

import re
from dataclasses import dataclass

MAX_ATTEMPTS = 3
BASE_DELAY_MS = 200
FACTOR = 2
JITTER_FRACTION = 0.5
TRANSIENT_STATUSES = [408, 429, 500, 502, 503, 504]

_BEARER_RE = re.compile(r"(bearer\s+)[A-Za-z0-9._-]+", re.IGNORECASE)


def is_retryable(*, idempotent: bool, status: int | None = None, is_timeout: bool = False) -> bool:
    """Retry only transient failures on idempotent calls (never mutating/non-idempotent)."""
    if not idempotent:
        return False
    if is_timeout:
        return True
    if status is None:
        return False
    return status in TRANSIENT_STATUSES


def base_delay_for_attempt(attempt: int) -> float:
    return BASE_DELAY_MS * (FACTOR ** attempt)


def backoff_schedule() -> list[float]:
    return [base_delay_for_attempt(a) for a in range(MAX_ATTEMPTS - 1)]


def redact_for_log(value: str) -> str:
    return _BEARER_RE.sub(r"\1<redacted>", value)


@dataclass
class InvokeResult:
    ok: bool
    attempts: int
    status: int | None = None
    body: object = None
    error: str | None = None
